src/libsyntax/errors/json.rs

```diff
@@ -173,37 +204,93 @@ impl<'a> Diagnostic<'a> {
                      .or_else(|| db.span.as_ref().map(|s| DiagnosticSpan::from_multispan(s, je)))
                      .unwrap_or(vec![]),
             children: vec![],
+            rendered: db.render_span.as_ref()
+                                    .and_then(|rsp| je.render(rsp)),
         }
     }
 }
 
 impl DiagnosticSpan {
+    fn from_span(span: Span, suggestion: Option<&String>, je: &JsonEmitter)
+                 -> DiagnosticSpan {
+        // obtain the full backtrace from the `macro_backtrace`
+        // helper; in some ways, it'd be better to expand the
+        // backtrace ourselves, but the `macro_backtrace` helper makes
+        // some decision, such as dropping some frames, and I don't
+        // want to duplicate that logic here.
+        let backtrace = je.cm.macro_backtrace(span).into_iter();
+        DiagnosticSpan::from_span_and_backtrace(span, suggestion, backtrace, je)
+    }
+
+    fn from_span_and_backtrace(span: Span,
+                               suggestion: Option<&String>,
+                               mut backtrace: vec::IntoIter<MacroBacktrace>,
+                               je: &JsonEmitter)
+                               -> DiagnosticSpan {
+        let start = je.cm.lookup_char_pos(span.lo);
+        let end = je.cm.lookup_char_pos(span.hi);
+        let backtrace_step =
+            backtrace.next()
+                     .map(|bt| {
+                         let call_site =
+                             Self::from_span_and_backtrace(bt.call_site,
+                                                           None,
+                                                           backtrace,
+                                                           je);
+                         let def_site_span =
+                             bt.def_site_span
+                               .map(|sp| {
+                                   Self::from_span_and_backtrace(sp,
+                                                                 None,
+                                                                 vec![].into_iter(),
+                                                                 je)
+                               });
+                         Box::new(DiagnosticSpanMacroExpansion {
+                             span: call_site,
+                             macro_decl_name: bt.macro_decl_name,
+                             def_site_span: def_site_span,
+                         })
+                     });
```

Comment:
Yeah I personally prefer avoiding rightward-drift as much as possible, the extra alignment here just seems a little confusing. This just looked the same as https://github.com/rust-lang-nursery/rustfmt/issues/439, but not a blocker at all!